lib/grandpa/grandpa.go

```diff
@@ -1297,3 +1209,192 @@ func (s *Service) lenVotes(stage Subround) int {
+		}
+		return err
+	}
+
+	err = s.blockState.SetFinalisedHash(msg.Vote.Hash, msg.Round, s.state.setID)
+	if err != nil {
+		return fmt.Errorf("setting finalised hash %s: %w", msg.Vote.Hash.Short(), err)
+	}
+
+	pcs, err := compactToJustification(msg.Precommits, msg.AuthData)
+	if err != nil {
+		return fmt.Errorf("compacting justification: %w", err)
+	}
+
+	err = s.grandpaState.SetPrecommits(msg.Round, msg.SetID, pcs)
+	if err != nil {
+		return fmt.Errorf("setting precommits: %w", err)
+	}
+
+	// ensure the persistence of informations
+	// and then notify the grandpa service about
+	// a received commit message
+	s.receivedCommit <- msg
+
+	// TODO: re-add catch-up logic (#1531)
+	return nil
+}
+
+func verifyCommitMessageJustification(fm *CommitMessage, setID uint64, threshold uint64,
+	authorities []*types.Authority, blockState BlockState) error {
+	if len(fm.Precommits) != len(fm.AuthData) {
+		return ErrPrecommitSignatureMismatch
+	}
+
+	if fm.SetID != setID {
+		return fmt.Errorf("%w: grandpa state set id %d, set id in the commit message %d",
+			ErrSetIDMismatch, setID, fm.SetID)
+	}
+
+	highestFinalizedHeader, err := blockState.GetHighestFinalisedHeader()
+	if err != nil {
+		return fmt.Errorf("getting highest finalised header: %w", err)
+	}
+
+	isDescendant, err := blockState.IsDescendantOf(highestFinalizedHeader.Hash(), fm.Vote.Hash)
+	if err != nil {
+		return fmt.Errorf("cannot verify ancestry of highest finalised block: %w", err)
+	}
+
+	if !isDescendant {
+		return errVoteBlockMismatch
+	}
+
+	eqvVoters := getEquivocatoryVoters(fm.AuthData)
+	var count int
+	for i, pc := range fm.Precommits {
+		just := &SignedVote{
+			Vote:        pc,
+			Signature:   fm.AuthData[i].Signature,
+			AuthorityID: fm.AuthData[i].AuthorityID,
+		}
+
+		err := verifyJustification(just, fm.Round, setID, precommit, authorities)
+		if err != nil {
+			logger.Errorf("failed to verify justification for vote from authority id %s, for block hash %s: %s",
+				just.AuthorityID.String(), just.Vote.Hash, err)
+			continue
+		}
+
+		isDescendant, err := blockState.IsDescendantOf(fm.Vote.Hash, just.Vote.Hash)
+		if err != nil {
+			logger.Warnf("could not check for descendant: %s", err)
+			continue
+		}
+
+		precommitedHeader, err := blockState.GetHeader(pc.Hash)
+		if err != nil {
+			return fmt.Errorf("getting header of %s: %w", pc.Hash.Short(), err)
+		}
+
+		if precommitedHeader.Number != uint(pc.Number) {
+			return fmt.Errorf("%w: expected number %d from header but got number %d",
+				ErrBlockHashMismatch, precommitedHeader.Number, pc.Number)
+		}
+
+		if _, ok := eqvVoters[fm.AuthData[i].AuthorityID]; ok {
+			continue
+		}
+
+		if isDescendant {
+			count++
+		}
+	}
+
+	// confirm total # signatures >= grandpa threshold
+	if uint64(count)+uint64(len(eqvVoters)) < threshold {
+		logger.Debugf(
+			"minimum votes not met for finalisation message. Need %d votes and received %d votes.",
+			threshold, count)
+		return ErrMinVotesNotMet
```

Comment:
just return an error with the context instead of debug logging it.

```suggestion
		return fmt.Errorf("%w: for finalisation message; need %d votes but received only %d votes",
			ErrMinVotesNotMet, threshold, uint64(count)+uint64(len(eqvVoters)))
```